// reakit/es/Radio/Radio.js
import 'reakit-utils/warning';
import { useCallback } from 'react';
import '../Id/IdProvider.js';
import { createComponent } from 'reakit-system/createComponent';
import { createHook } from 'reakit-system/createHook';
import '../Box/Box.js';
import { _ as _objectWithoutPropertiesLoose, a as _objectSpread2 } from '../_rollupPluginBabelHelpers-9b9ba5c9.js';
import 'reakit-utils/mergeRefs';
import 'reakit-utils/tabbable';
import 'reakit-utils/hasFocusWithin';
import '../Tabbable/Tabbable.js';
import { useAllCallbacks } from 'reakit-utils/useAllCallbacks';
import 'reakit-utils/useSealedState';
import '../Id/IdState.js';
import 'reakit-utils/createOnKeyDown';
import '../Id/Id.js';
import '../Rover/RoverState.js';
import { useRover } from '../Rover/Rover.js';
import { useRadioState } from './RadioState.js';

var useRadio = createHook({
  name: "Radio",
  compose: useRover,
  useState: useRadioState,
  keys: ["value", "checked"],
  useOptions: function useOptions(_ref) {
    var _ref$unstable_clickOn = _ref.unstable_clickOnEnter,
        unstable_clickOnEnter = _ref$unstable_clickOn === void 0 ? false : _ref$unstable_clickOn,
        options = _objectWithoutPropertiesLoose(_ref, ["unstable_clickOnEnter"]);

    return _objectSpread2({
      unstable_clickOnEnter: unstable_clickOnEnter
    }, options);
  },
  useProps: function useProps(options, _ref2) {
    var htmlOnChange = _ref2.onChange,
        htmlOnClick = _ref2.onClick,
        htmlProps = _objectWithoutPropertiesLoose(_ref2, ["onChange", "onClick"]);

    var checked = typeof options.checked !== "undefined" ? options.checked : options.state === options.value;
    var onChange = useCallback(function (event) {
      if (htmlOnChange) {
        htmlOnChange(event);
      }

      if (options.disabled || !options.setState) return;
      options.setState(options.value);
    }, [htmlOnChange, options.disabled, options.setState, options.value]);
    var onClick = useCallback(function (event) {
      if (event.target instanceof HTMLInputElement) return;
      onChange(event);
    }, [onChange]);
    return _objectSpread2({
      checked: checked,
      "aria-checked": checked,
      value: options.value,
      role: "radio",
      type: "radio",
      onChange: onChange,
      onClick: useAllCallbacks(onClick, htmlOnClick)
    }, htmlProps);
  }
});
var Radio = createComponent({
  as: "input",
  useHook: useRadio
});

export { Radio, useRadio };
