// reakit/es/Form/FormRadioGroup.js
import 'reakit-utils/warning';
import { createContext, useMemo, useCallback, createElement } from 'react';
import 'reakit-utils/useId';
import '../Id/IdProvider.js';
import { createComponent } from 'reakit-system/createComponent';
import { createHook } from 'reakit-system/createHook';
import '../Box/Box.js';
import { _ as _objectWithoutPropertiesLoose, a as _objectSpread2 } from '../_rollupPluginBabelHelpers-9b9ba5c9.js';
import 'reakit-utils/useSealedState';
import 'reakit-utils/useLiveRef';
import { usePipe } from 'reakit-utils/usePipe';
import '../Id/IdState.js';
import 'reakit-utils/useUpdateEffect';
import 'reakit-utils/isPromise';
import 'reakit-utils/isEmpty';
import 'reakit-utils/isObject';
import { unstable_useFormState } from './FormState.js';
import './utils/setAllIn.js';
import './utils/getIn.js';
import 'reakit-utils/toArray';
import 'reakit-utils/isInteger';
import './utils/setIn.js';
import '../formatInputName-56bda968.js';
import '../getInputId-0a89895d.js';
import '../getLabelId-f0ada44c.js';
import '../shouldShowError-c65f557e.js';
import '../Group/Group.js';
import { unstable_useFormGroup } from './FormGroup.js';
import { useRoverState } from '../Rover/RoverState.js';

var FormRadioGroupContext = createContext(null);
var unstable_useFormRadioGroup = createHook({
  name: "FormRadioGroup",
  compose: unstable_useFormGroup,
  useState: unstable_useFormState,
  keys: ["name"],
  useProps: function useProps(_, _ref) {
    var htmlWrap = _ref.unstable_wrap,
        htmlProps = _objectWithoutPropertiesLoose(_ref, ["unstable_wrap"]);

    var rover = useRoverState({
      loop: true
    });
    var providerValue = useMemo(function () {
      return rover;
    }, [rover.stops, rover.currentId, rover.unstable_pastId]);
    var wrap = useCallback(function (children) {
      return createElement(FormRadioGroupContext.Provider, {
        value: providerValue
      }, children);
    }, [providerValue]);
    return _objectSpread2({
      role: "radiogroup",
      unstable_wrap: usePipe(wrap, htmlWrap)
    }, htmlProps);
  }
});
var unstable_FormRadioGroup = createComponent({
  as: "fieldset",
  useHook: unstable_useFormRadioGroup
});

export { FormRadioGroupContext, unstable_FormRadioGroup, unstable_useFormRadioGroup };
