// reakit/es/Form/FormMessage.js
import 'react';
import 'reakit-utils/useId';
import { createComponent } from 'reakit-system/createComponent';
import { createHook } from 'reakit-system/createHook';
import { useBox } from '../Box/Box.js';
import { a as _objectSpread2 } from '../_rollupPluginBabelHelpers-9b9ba5c9.js';
import 'reakit-utils/useSealedState';
import 'reakit-utils/useLiveRef';
import 'reakit-utils/useUpdateEffect';
import 'reakit-utils/isPromise';
import 'reakit-utils/isEmpty';
import 'reakit-utils/isObject';
import { unstable_useFormState } from './FormState.js';
import './utils/setAllIn.js';
import { unstable_getIn } from './utils/getIn.js';
import 'reakit-utils/toArray';
import 'reakit-utils/isInteger';
import './utils/setIn.js';
import '../formatInputName-56bda968.js';
import '../getInputId-0a89895d.js';
import { s as shouldShowError, g as getMessageId } from '../shouldShowError-c65f557e.js';

function shouldShowMessage(_ref, name) {
  var touched = _ref.touched,
      messages = _ref.messages;
  return Boolean(unstable_getIn(touched, name) && unstable_getIn(messages, name));
}

var unstable_useFormMessage = createHook({
  name: "FormMessage",
  compose: useBox,
  useState: unstable_useFormState,
  keys: ["name"],
  useProps: function useProps(options, htmlProps) {
    var children = shouldShowError(options, options.name) ? unstable_getIn(options.errors, options.name) : undefined;
    children = children || (shouldShowMessage(options, options.name) ? unstable_getIn(options.messages, options.name) : undefined);
    return _objectSpread2({
      role: "alert",
      id: getMessageId(options.name, options.baseId),
      children: children
    }, htmlProps);
  }
});
var unstable_FormMessage = createComponent({
  as: "div",
  useHook: unstable_useFormMessage
});

export { unstable_FormMessage, unstable_useFormMessage };
