// reakit/lib/Form/FormRadioGroup.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('reakit-utils/warning');
var React = require('react');
require('reakit-utils/useId');
require('../Id/IdProvider.js');
var createComponent = require('reakit-system/createComponent');
var createHook = require('reakit-system/createHook');
require('../Box/Box.js');
var _rollupPluginBabelHelpers = require('../_rollupPluginBabelHelpers-a1eb266e.js');
require('reakit-utils/useSealedState');
require('reakit-utils/useLiveRef');
var usePipe = require('reakit-utils/usePipe');
require('../Id/IdState.js');
require('reakit-utils/useUpdateEffect');
require('reakit-utils/isPromise');
require('reakit-utils/isEmpty');
require('reakit-utils/isObject');
var FormState = require('./FormState.js');
require('./utils/setAllIn.js');
require('./utils/getIn.js');
require('reakit-utils/toArray');
require('reakit-utils/isInteger');
require('./utils/setIn.js');
require('../formatInputName-96726777.js');
require('../getInputId-cebd1149.js');
require('../getLabelId-e0169cb0.js');
require('../shouldShowError-db4a6670.js');
require('../Group/Group.js');
var FormGroup = require('./FormGroup.js');
var RoverState = require('../Rover/RoverState.js');

var FormRadioGroupContext = React.createContext(null);
var unstable_useFormRadioGroup = createHook.createHook({
  name: "FormRadioGroup",
  compose: FormGroup.unstable_useFormGroup,
  useState: FormState.unstable_useFormState,
  keys: ["name"],
  useProps: function useProps(_, _ref) {
    var htmlWrap = _ref.unstable_wrap,
        htmlProps = _rollupPluginBabelHelpers._objectWithoutPropertiesLoose(_ref, ["unstable_wrap"]);

    var rover = RoverState.useRoverState({
      loop: true
    });
    var providerValue = React.useMemo(function () {
      return rover;
    }, [rover.stops, rover.currentId, rover.unstable_pastId]);
    var wrap = React.useCallback(function (children) {
      return React.createElement(FormRadioGroupContext.Provider, {
        value: providerValue
      }, children);
    }, [providerValue]);
    return _rollupPluginBabelHelpers._objectSpread2({
      role: "radiogroup",
      unstable_wrap: usePipe.usePipe(wrap, htmlWrap)
    }, htmlProps);
  }
});
var unstable_FormRadioGroup = createComponent.createComponent({
  as: "fieldset",
  useHook: unstable_useFormRadioGroup
});

exports.FormRadioGroupContext = FormRadioGroupContext;
exports.unstable_FormRadioGroup = unstable_FormRadioGroup;
exports.unstable_useFormRadioGroup = unstable_useFormRadioGroup;
