// reakit/lib/Form/FormCheckbox.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('reakit-utils/warning');
var React = require('react');
require('reakit-utils/useId');
var createComponent = require('reakit-system/createComponent');
var createHook = require('reakit-system/createHook');
require('../Box/Box.js');
var _rollupPluginBabelHelpers = require('../_rollupPluginBabelHelpers-a1eb266e.js');
require('reakit-utils/mergeRefs');
require('reakit-utils/tabbable');
require('reakit-utils/hasFocusWithin');
require('../Tabbable/Tabbable.js');
require('reakit-utils/removeIndexFromArray');
var useAllCallbacks = require('reakit-utils/useAllCallbacks');
require('reakit-utils/useSealedState');
require('../Checkbox/CheckboxState.js');
var Checkbox = require('../Checkbox/Checkbox.js');
require('reakit-utils/useLiveRef');
require('reakit-utils/useUpdateEffect');
require('reakit-utils/isPromise');
require('reakit-utils/isEmpty');
require('reakit-utils/isObject');
var FormState = require('./FormState.js');
require('./utils/setAllIn.js');
var getIn = require('./utils/getIn.js');
require('reakit-utils/toArray');
require('reakit-utils/isInteger');
require('./utils/setIn.js');
var formatInputName = require('../formatInputName-96726777.js');
var getInputId = require('../getInputId-cebd1149.js');
var getLabelId = require('../getLabelId-e0169cb0.js');
var shouldShowError = require('../shouldShowError-db4a6670.js');

var unstable_useFormCheckbox = createHook.createHook({
  name: "FormCheckbox",
  compose: Checkbox.useCheckbox,
  useState: FormState.unstable_useFormState,
  keys: ["name", "value"],
  useOptions: function useOptions(options) {
    var state = getIn.unstable_getIn(options.values, options.name);

    var setState = function setState(value) {
      return options.update(options.name, value);
    };

    return _rollupPluginBabelHelpers._objectSpread2({}, options, {
      state: state,
      setState: setState
    });
  },
  useProps: function useProps(options, _ref) {
    var htmlOnBlur = _ref.onBlur,
        htmlProps = _rollupPluginBabelHelpers._objectWithoutPropertiesLoose(_ref, ["onBlur"]);

    var isBoolean = typeof options.value === "undefined";
    var onBlur = React.useCallback(function () {
      options.blur(options.name);
    }, [options.blur, options.name]);
    return _rollupPluginBabelHelpers._objectSpread2({
      "aria-invalid": shouldShowError.shouldShowError(options, options.name),
      name: formatInputName.formatInputName(options.name),
      onBlur: useAllCallbacks.useAllCallbacks(onBlur, htmlOnBlur)
    }, isBoolean ? {
      id: getInputId.getInputId(options.name, options.baseId),
      "aria-describedby": shouldShowError.getMessageId(options.name, options.baseId),
      "aria-labelledby": getLabelId.getLabelId(options.name, options.baseId)
    } : {}, {}, htmlProps);
  }
});
var unstable_FormCheckbox = createComponent.createComponent({
  as: "input",
  useHook: unstable_useFormCheckbox
});

exports.unstable_FormCheckbox = unstable_FormCheckbox;
exports.unstable_useFormCheckbox = unstable_useFormCheckbox;
