// reakit/lib/Tab/TabState.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('reakit-utils/warning');
var React = require('react');
require('../Id/IdProvider.js');
var _rollupPluginBabelHelpers = require('../_rollupPluginBabelHelpers-a1eb266e.js');
var useSealedState = require('reakit-utils/useSealedState');
require('../Id/IdState.js');
var RoverState = require('../Rover/RoverState.js');

function useTabState(initialState) {
  if (initialState === void 0) {
    initialState = {};
  }

  var _useSealedState = useSealedState.useSealedState(initialState),
      _useSealedState$selec = _useSealedState.selectedId,
      sealedSelectedId = _useSealedState$selec === void 0 ? null : _useSealedState$selec,
      _useSealedState$loop = _useSealedState.loop,
      loop = _useSealedState$loop === void 0 ? true : _useSealedState$loop,
      _useSealedState$manua = _useSealedState.manual,
      manual = _useSealedState$manua === void 0 ? false : _useSealedState$manua,
      sealed = _rollupPluginBabelHelpers._objectWithoutPropertiesLoose(_useSealedState, ["selectedId", "loop", "manual"]);

  var _React$useState = React.useState(sealedSelectedId),
      selectedId = _React$useState[0],
      select = _React$useState[1];

  var rover = RoverState.useRoverState(_rollupPluginBabelHelpers._objectSpread2({
    loop: loop,
    currentId: selectedId
  }, sealed));
  return _rollupPluginBabelHelpers._objectSpread2({}, rover, {
    selectedId: selectedId,
    manual: manual,
    select: select
  });
}
var keys = [].concat(RoverState.useRoverState.__keys, ["selectedId", "select", "manual"]);
useTabState.__keys = keys;

exports.useTabState = useTabState;
